[torch] make StreamContext nop (and doesn't call cuda lazy init) when stream is None

jspark1105 opened the pull request:
Summary: Make it really nop.

Test Plan: buck run mode/opt -c fbcode.nvcc_arch=v100 //hpc/models/ads:ads_10x_launcher +launcher=local +data_loader=random +mode=local_cpu optimizer.sparse_optim=ROWWISE_ADAGRAD data_loader.batch_size=16384 launcher.inactive_timeout_sec=null model.jit=True model.ne_opt_level=1

Differential Revision: D27433887

nikithamalgifb commented:
Can you try running `python test/test_jit.py -v TestCUDA` and see if it doesn't break anything in JIT?

nikithamalgifb commented on torch/cuda/__init__.py:
```diff
@@ -339,8 +339,8 @@ class StreamContext(object):
-        self.src_prev_stream = torch.cuda.default_stream(None)
-        self.dst_prev_stream = torch.cuda.default_stream(None)
+        self.src_prev_stream = torch.cuda.default_stream(None) if stream else None
+        self.dst_prev_stream = torch.cuda.default_stream(None) if stream else None
```
Can you try this? 

self.src_prev_stream = torch.cuda.default_stream(None)
self.dst_prev_stream = torch.cuda.default_stream(None)
        if not torch.jit.is_scripting():
            self.src_prev_stream = None
            self.dst_prev_stream = None
See if this works for the buck test?

nikithamalgifb replied:
Tried this fix and it works locally. I couldn't run the buck test.

jspark1105 committed:
[torch] make StreamContext nop (and doesn't call cuda lazy init) when stream is None (#54963)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/54963

Make it really nop.

Test Plan: buck run mode/opt -c fbcode.nvcc_arch=v100 //hpc/models/ads:ads_10x_launcher +launcher=local +data_loader=random +mode=local_cpu optimizer.sparse_optim=ROWWISE_ADAGRAD data_loader.batch_size=16384 launcher.inactive_timeout_sec=null model.jit=True model.ne_opt_level=1

Differential Revision: D27433887

fbshipit-source-id: 24ecb5723de2ed4176b5c69d69a2589c31b67618
